fineract-provider/src/main/java/org/apache/fineract/portfolio/client/domain/AccountNumberGenerator.java

```diff
@@ -134,11 +147,50 @@ private String generateAccountNumber(Map<String, String> propertyMap, AccountNum
                 prefix = prefix.substring(0, Math.min(prefix.length(), 10));
             }
 
+            if (accountNumberPrefixType.getValue() == 401) {
+
+                AccountNumberFormat accountNumberByPrefix = this.accountNumberFormatRepository
+                        .findOneByPrefixEnumAccountTypeEnum(accountNumberFormat.getAccountTypeEnum(), 401);
+                Integer digits = accountNumberByPrefix.getDigits();
+                Integer numberLength = accountMaxLength - digits;
+                accountNumber = StringUtils.leftPad(propertyMap.get(ID), numberLength, '0');
+
+            } else {
+                accountNumber = StringUtils.leftPad(accountNumber, Integer.valueOf(propertyMap.get(ID).length()), '0');
+            }
+
             accountNumber = StringUtils.overlay(accountNumber, prefix, 0, 0);
         }
         return accountNumber;
     }
 
+    private Map<String, String> prefixEight(Map<String, String> propertyMap, String accountNumber, Integer accountMaxLength,
+            AccountNumberFormat accountNumberFormat) {
+        AccountNumberFormat accountNumberByPrefix = this.accountNumberFormatRepository
+                .findOneByPrefixEnumAccountTypeEnum(accountNumberFormat.getAccountTypeEnum(), 401);
+        Integer digits = accountNumberByPrefix.getDigits();
+        String prefix = accountNumberByPrefix.getPrefixCharacter();
+
+        Integer numberLength = accountMaxLength - digits;
+
+        if (Integer.valueOf(propertyMap.get(ID).length()) == numberLength + 1) { // 888-999999
+            prefix = prefix.substring(0, prefix.length() - 1);
```

Comment:
This logic seems to be a) hardcoded to expect a specific length for prefix and digits b) only handle specific cases of length/prefix length. Can this not be made generic so that we automatically work out the length of the number / prefix based on the lengths of the inputs?